packages/flutter_tools/lib/src/android/android_workflow.dart

```diff
@@ -426,4 +425,32 @@ class AndroidLicenseValidator extends DoctorValidator {
     }
     return _processManager.canRun(sdkManagerPath);
   }
+
+  String messageForSdkManagerError(
+    Java java,
+    AndroidSdk androidSdk,
+    List<String> androidSdkStderr,
+    int exitCode,
+    Platform platform,
+  ) {
+    final String sdkManagerPath = androidSdk.sdkManagerPath ?? '';
+
+    final bool failedDueToJdkIncompatibility = androidSdkStderr.join().contains(
+      RegExp(r'.*java\.lang\.UnsupportedClassVersionError.*SdkManagerCli '
+        r'has been compiled by a more recent version of the Java Runtime.*'));
+
+    if (failedDueToJdkIncompatibility) {
+      return 'Android sdkmanager tool was found, but failed to run ($sdkManagerPath): "exited code $exitCode".\n'
+        'It appears the version of the Java binary used (${java.binaryPath}) is '
+        'too out-of-date and is incompatible with the Android sdkmanager tool.\n'
+        'If the Java binary came bundled with Android Studio, consider updating '
+        'your installation of Android studio. Alternatively, you can uninstall '
+        'the Android SDK command-line tools and install an earlier version. ';
+    }
+    throwToolExit(_userMessages.androidCannotRunSdkManager(
```

Comment:
i think this function is overloaded with several tasks, which are not obvious from its name. I think `messageForSdkManagerError()` should just be responsible for interpolating values into the message string. I would recommend lifting the other logic into the call site around line 409.

Reply:
Ah, I'm sorry. Originally, I went back and forth between having this function throw a tool exit or  just generating the message and letting the caller throw. Looks like I left this in an in-between state.

Reply:
> Ah, I'm sorry the headache reading this code probably caused.

it's all good. i've submitted worse that i've marked "ready for review"

Reply:
I've changed this function to not throw a `ToolExit` (it just returns the error message). Commit: [make messageForSdkManagerError private](https://github.com/flutter/flutter/pull/138762/commits/face2518c1d652d0aaaf2d9b0f7dacd14d29efea)

> `messageForSdkManagerError()` should just be responsible for interpolating values into the message string. I would recommend lifting the other logic into the call site around line 409.

If by "other logic" you mean the conditional logic to choose what string to build, I'm not sure I agree. `runLicense` manager already has a decent length and amount of conditional logic going on, and I don't want to add another layer of depth (which also has a bunch of noisy string interpolation) to it. I'm happy to reconsider if you can clarify or make a code suggestion

Reply:
this lgtm